File: backend/app/schemas/processing.py
from datetime import datetime
from typing import Literal
import uuid

from pydantic import BaseModel, Field

ProcessingStage = Literal["article", "classification", "ioc", "tagging"]
ProcessingState = Literal["pending", "queued", "running", "retry_wait", "attention"]


class ProcessingWorkResponse(BaseModel):
    item_id: uuid.UUID
    stage: ProcessingStage
    revision: str
    title: str
    feed_id: uuid.UUID
    feed_name: str
    state: ProcessingState
    reason: str | None
    message: str | None
    first_seen_at: datetime
    age_seconds: int
    attempts: int
    next_retry_at: datetime | None
    can_retry: bool


class ProcessingWorkList(BaseModel):
    items: list[ProcessingWorkResponse]
    next_cursor: str | None
    has_more: bool


class ProcessingSelection(BaseModel):
    item_id: uuid.UUID
    stage: ProcessingStage
    revision: str = Field(min_length=1, max_length=64)


class ProcessingRecoveryRequest(BaseModel):
    idempotency_key: uuid.UUID
    items: list[ProcessingSelection] = Field(min_length=1, max_length=100)


class ProcessingRecoveryCancel(BaseModel):
    expected_version: int = Field(ge=1)


class ProcessingRecoveryItemResponse(BaseModel):
    item_id: uuid.UUID
    stage: ProcessingStage
    state: Literal["queued", "running", "succeeded", "failed", "cancelled"]
    reason: str | None
    message: str | None
    title: str | None
    feed_name: str | None


class ProcessingRecoveryResponse(BaseModel):
    id: uuid.UUID
    version: int
    status: Literal["queued", "running", "succeeded", "partial", "cancelled", "failed"]
    created_at: datetime
    updated_at: datetime
    total_count: int
    completed_count: int
    failed_count: int
    cancelled_count: int
    can_cancel: bool
    access_limited: bool = False
    items: list[ProcessingRecoveryItemResponse]


class ProcessingRecoveryList(BaseModel):
    items: list[ProcessingRecoveryResponse]
    next_cursor: str | None
    has_more: bool

File: web/src/types/processing.ts
export type ProcessingStage = 'article' | 'classification' | 'ioc' | 'tagging'
export type ProcessingWorkState = 'pending' | 'queued' | 'running' | 'retry_wait' | 'attention'
export type ProcessingRunStatus = 'queued' | 'running' | 'succeeded' | 'partial' | 'cancelled' | 'failed'

export interface ProcessingWorkIdentity {
  item_id: string
  stage: ProcessingStage
  revision: string
}

export interface ProcessingWork extends ProcessingWorkIdentity {
  title: string
  feed_id: string
  feed_name: string
  state: ProcessingWorkState
  reason: string | null
  message: string | null
  first_seen_at: string
  age_seconds: number
  attempts: number
  next_retry_at: string | null
  can_retry: boolean
}

export interface ProcessingPage<T> {
  items: T[]
  next_cursor: string | null
  has_more: boolean
}

export interface ProcessingRecoveryRequest {
  idempotency_key: string
  items: ProcessingWorkIdentity[]
}

export interface ProcessingRunItem {
  item_id: string
  stage: ProcessingStage
  state: 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled'
  title: string | null
  feed_name: string | null
  reason: string | null
  message: string | null
}

export interface ProcessingRun {
  id: string
  version: number
  status: ProcessingRunStatus
  created_at: string
  updated_at: string
  total_count: number
  completed_count: number
  failed_count: number
  cancelled_count: number
  can_cancel: boolean
  access_limited?: boolean
  items: ProcessingRunItem[]
}
